File: Web/packages/kernel/test/net-conformance.test.ts
//
//  net-conformance.test.ts — the SHARED connectivity corpus through the TS kernel
//  (OpenSource/Conformance/net/{status,transitions}.json). The Kotlin twin (:core
//  NetConformanceTest) and the Swift reference (NetConformance, record lane) run the SAME
//  files, so a captive portal, a personal hotspot and a flapping interface cannot mean one
//  thing on one renderer and something else on another.
//
//  Missing corpus = loud failure, and so is an empty section: a silently-skipped conformance
//  suite is how drift starts, and a runner that reads three of five case arrays is the same
//  defect in a new place.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import {
  classifyPath, isOnline, radioGeneration, probeReachable, NetDebounce, type NetSnapshot,
} from "../src/net-core.ts";

type Json = { [key: string]: any };

function corpusDir(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/net");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("OpenSource/Conformance/net not found");
    dir = parent;
  }
}

function corpus(file: string): Json {
  const doc = JSON.parse(readFileSync(join(corpusDir(), file), "utf-8")) as Json;
  assert.equal(doc["version"], 1, `${file}: version`);
  return doc;
}

function section(doc: Json, name: string, file: string): Json[] {
  const raw = doc[name];
  const rows = Array.isArray(raw) ? raw : (raw as Json | undefined)?.["cases"];
  assert.ok(Array.isArray(rows) && rows.length > 0, `${file}: ${name} must be a non-empty case array`);
  return rows as Json[];
}

test("net: the classification fold agrees with the corpus", () => {
  const doc = corpus("status.json");
  for (const testCase of section(doc, "classify", "status.json")) {
    const given = testCase["given"] as Json;
    const expect = testCase["expect"] as Json;
    const actual = classifyPath({
      satisfied: given["satisfied"] === true,
      interfaces: given["interfaces"] as string[],
      transports: given["transports"] as string[],
      metered: given["metered"] === true,
      dataSaver: given["dataSaver"] === true,
    });
    assert.equal(actual.reachable, expect["reachable"], `${testCase["name"]}: reachable`);
    assert.equal(actual.type, expect["type"], `${testCase["name"]}: type`);
    assert.equal(actual.expensive, expect["expensive"], `${testCase["name"]}: expensive`);
    assert.equal(actual.constrained, expect["constrained"], `${testCase["name"]}: constrained`);
  }
});

test("net: the online split agrees with the corpus", () => {
  const doc = corpus("status.json");
  for (const testCase of section(doc, "online", "status.json")) {
    const given = testCase["given"] as Json;
    const snapshot: NetSnapshot = {
      reachable: given["reachable"] === true,
      type: given["reachable"] === true ? "wifi" : "none",
      expensive: false,
      constrained: false,
      validated: true,
    };
    assert.equal(
      isOnline(snapshot, given["probeFailed"] === true),
      (testCase["expect"] as Json)["online"],
      `${testCase["name"]}: online`,
    );
  }
});

test("net: the radio-family map agrees with the corpus", () => {
  const doc = corpus("status.json");
  for (const testCase of section(doc, "generation", "status.json")) {
    const given = testCase["given"] as Json;
    assert.equal(
      radioGeneration(given["type"] as string, given["radio"] as string | null),
      (testCase["expect"] as Json)["generation"],
      `${testCase["name"]}: generation`,
    );
  }
});

test("net: the probe verdict agrees with the corpus", () => {
  const doc = corpus("status.json");
  for (const testCase of section(doc, "probe", "status.json")) {
    const given = testCase["given"] as Json;
    assert.equal(
      probeReachable(given["status"] as number, given["requestHost"] as string, given["location"] as string | null),
      (testCase["expect"] as Json)["reachable"],
      `${testCase["name"]}: reachable`,
    );
  }
});

test("net: the debounce timelines agree with the corpus", () => {
  const doc = corpus("transitions.json");
  const debounceMs = doc["debounceMs"] as number;
  assert.equal(typeof debounceMs, "number", "transitions.json: debounceMs");

  for (const timeline of section(doc, "timelines", "transitions.json")) {
    const name = timeline["name"] as string;
    const steps = timeline["steps"] as Json[];
    assert.ok(Array.isArray(steps) && steps.length > 0, `${name}: no steps`);
    const machine = new NetDebounce(debounceMs);
    const emitted: Json[] = [];

    for (const step of steps) {
      const at = step["at"] as number;
      machine.advance(at);
      if (step["path"] !== undefined && step["path"] !== null) {
        const path = step["path"] as Json;
        machine.path(at, {
          reachable: path["reachable"] === true,
          type: path["type"] as string,
          expensive: path["expensive"] === true,
          constrained: path["constrained"] === true,
          validated: true,
        });
      } else if (step["probeFailed"] !== undefined) {
        machine.probe(at, step["probeFailed"] === true);
      }
      for (const change of machine.drain()) {
        emitted.push({
          at: change.at,
          event: "change",
          data: {
            online: change.online,
            reachable: change.snapshot.reachable,
            type: change.snapshot.type,
            expensive: change.snapshot.expensive,
            constrained: change.snapshot.constrained,
            previous: change.previous,
          },
        });
      }
    }

    const expect = timeline["expect"] as Json;
    assert.deepEqual(emitted, expect["events"], `${name}: events`);
    const context = expect["context"] as Json;
    assert.equal(machine.online, context["online"], `${name}: context.online`);
    assert.equal(machine.settled.type, context["type"], `${name}: context.type`);
    assert.equal(machine.settled.expensive, context["expensive"], `${name}: context.expensive`);
    assert.equal(machine.settled.constrained, context["constrained"], `${name}: context.constrained`);
  }
});

File: Engine/Android/core/src/test/kotlin/despia/engine/NetConformanceTest.kt
package despia.engine

import java.io.File
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The connectivity conformance runner - executes
 * OpenSource/Conformance/net/{status,transitions}.json through THIS runtime's NetCore fold and
 * NetDebounce machine (parity/F05-net.md). The TS twin (@despia/kernel net-core.ts,
 * net-conformance.test.ts) and the Swift reference (NetConformance) run the SAME files, so a
 * captive portal, a personal hotspot and a flapping interface cannot mean one thing on one
 * renderer and something else on another.
 *
 * Missing corpus = loud failure, and so is an empty section: a runner that reads three of five
 * case arrays is the same defect as a corpus nobody runs.
 */
class NetConformanceTest {

    private fun corpusFile(name: String): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/net/$name")
            if (candidate.isFile) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/net/$name not found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun root(name: String): Map<String, Any?> {
        val root = json(corpusFile(name).readText()).foundationValue as? Map<String, Any?>
            ?: error("$name: not a JSON object")
        assertEquals(1, (root["version"] as? Number)?.toInt(), "$name version")
        return root
    }

    @Suppress("UNCHECKED_CAST")
    private fun section(root: Map<String, Any?>, name: String, file: String): List<Map<String, Any?>> {
        val raw = root[name]
        val rows = (raw as? List<Any?>) ?: ((raw as? Map<String, Any?>)?.get("cases") as? List<Any?>)
        val cases = rows?.map { it as Map<String, Any?> }
        assertTrue(cases != null && cases.isNotEmpty(), "$file: $name must be a non-empty case array")
        return cases!!
    }

    private fun bool(value: Any?): Boolean = value == true

    @Test
    @Suppress("UNCHECKED_CAST")
    fun classifyAgreesWithCorpus() {
        for (case in section(root("status.json"), "classify", "status.json")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val given = case["given"] as Map<String, Any?>
            val expect = case["expect"] as Map<String, Any?>
            val actual = NetCore.classify(
                satisfied = bool(given["satisfied"]),
                interfaces = (given["interfaces"] as? List<Any?>)?.map { it as String } ?: emptyList(),
                transports = (given["transports"] as? List<Any?>)?.map { it as String } ?: emptyList(),
                metered = bool(given["metered"]),
                dataSaver = bool(given["dataSaver"]),
            )
            assertEquals(expect["reachable"], actual.reachable, "$name: reachable")
            assertEquals(expect["type"], actual.type, "$name: type")
            assertEquals(expect["expensive"], actual.expensive, "$name: expensive")
            assertEquals(expect["constrained"], actual.constrained, "$name: constrained")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun onlineSplitAgreesWithCorpus() {
        for (case in section(root("status.json"), "online", "status.json")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val given = case["given"] as Map<String, Any?>
            val reachable = bool(given["reachable"])
            val snapshot = NetSnapshot(reachable, if (reachable) "wifi" else "none", false, false)
            assertEquals(
                (case["expect"] as Map<String, Any?>)["online"],
                NetCore.online(snapshot, bool(given["probeFailed"])),
                "$name: online",
            )
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun generationAgreesWithCorpus() {
        for (case in section(root("status.json"), "generation", "status.json")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val given = case["given"] as Map<String, Any?>
            assertEquals(
                (case["expect"] as Map<String, Any?>)["generation"],
                NetCore.generation(given["type"] as String, given["radio"] as? String),
                "$name: generation",
            )
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun probeVerdictAgreesWithCorpus() {
        for (case in section(root("status.json"), "probe", "status.json")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val given = case["given"] as Map<String, Any?>
            assertEquals(
                (case["expect"] as Map<String, Any?>)["reachable"],
                NetCore.probeReachable(
                    (given["status"] as Number).toInt(),
                    given["requestHost"] as? String,
                    given["location"] as? String,
                ),
                "$name: reachable",
            )
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun debounceTimelinesAgreeWithCorpus() {
        val doc = root("transitions.json")
        val debounceMs = (doc["debounceMs"] as? Number)?.toLong() ?: error("transitions.json: no debounceMs")

        for (timeline in section(doc, "timelines", "transitions.json")) {
            val name = timeline["name"] as? String ?: "<unnamed>"
            val steps = (timeline["steps"] as? List<Any?>)?.map { it as Map<String, Any?> }
                ?: error("$name: no steps")
            assertTrue(steps.isNotEmpty(), "$name: no steps")

            val machine = NetDebounce(debounceMs)
            val emitted = ArrayList<Map<String, Any?>>()

            for (step in steps) {
                val at = (step["at"] as Number).toLong()
                machine.advance(at)
                val path = step["path"] as? Map<String, Any?>
                if (path != null) {
                    machine.path(at, NetSnapshot(
                        reachable = bool(path["reachable"]),
                        type = path["type"] as String,
                        expensive = bool(path["expensive"]),
                        constrained = bool(path["constrained"]),
                    ))
                } else if (step.containsKey("probeFailed")) {
                    machine.probe(at, bool(step["probeFailed"]))
                }
                for (change in machine.drain()) {
                    emitted.add(mapOf(
                        "at" to change.at,
                        "event" to "change",
                        "data" to mapOf(
                            "online" to change.online,
                            "reachable" to change.snapshot.reachable,
                            "type" to change.snapshot.type,
                            "expensive" to change.snapshot.expensive,
                            "constrained" to change.snapshot.constrained,
                            "previous" to change.previous,
                        ),
                    ))
                }
            }

            val expect = timeline["expect"] as Map<String, Any?>
            val expectedEvents = (expect["events"] as List<Any?>).map { row ->
                val event = row as Map<String, Any?>
                val data = event["data"] as Map<String, Any?>
                mapOf(
                    "at" to (event["at"] as Number).toLong(),
                    "event" to event["event"],
                    "data" to mapOf(
                        "online" to data["online"],
                        "reachable" to data["reachable"],
                        "type" to data["type"],
                        "expensive" to data["expensive"],
                        "constrained" to data["constrained"],
                        "previous" to data["previous"],
                    ),
                )
            }
            assertEquals(expectedEvents, emitted.toList(), "$name: events")

            val context = expect["context"] as Map<String, Any?>
            assertEquals(context["online"], machine.online, "$name: context.online")
            assertEquals(context["type"], machine.settled.type, "$name: context.type")
            assertEquals(context["expensive"], machine.settled.expensive, "$name: context.expensive")
            assertEquals(context["constrained"], machine.settled.constrained, "$name: context.constrained")
        }
    }
}
